### PR TITLE
Simplify return convention of verify_ functions

### PR DESCRIPTION
**Pull request by asn-d6**
See https://github.com/ethereum/c-kzg-4844/issues/139

### CODE REVIEW

**Comment by asn-d6** [comment]
This is still a draft. Haven't started adapting the bindings yet.

**Review by jtraglia — APPROVED** [review]
Looks good! Just a couple of small nits.

File: src/c_kzg_4844.h
Comment (ppopth) [reply src/c_kzg_4844.h:100]: I think `BAD_ARGS` is a good idea. However, please note that removing or reordering makes the underlying integers change which will probably make the new version of c-kzg-4844 not backward-compatible.

Comment (asn-d6) [reply src/c_kzg_4844.h:100]: I agree this change is a good idea; I tried to do it but it actually bloats the PR in a non-trivial way. I suggest we open another issue for this and do it at a later stage, so that it doesn't delay the review of this PR.

(currently working on the bindings)

**Comment by asn-d6** [comment]
Pushed some more work here.

I have a question about the Java bindings (/cc @StefanBratanov @benjaminion):

The java bindings right now, upon receiving a bad KZG proof, might either throw an exception (when we go OOM, or when a group element is not canonical) or return false (if the verification equation fails). This is a bit weird because an adversary can send a non-canonical group element or a non-verifying proof with the same ease, and they will be handled differently.

With this PR we can simplify this logic. What would you prefer to do here? Here are some options:

1) Make all the verification functions have no retval. If they finish without any exceptions, then the proof is totally valid. Exceptions can be fine-grained but will be handled by clients.

2) Throw exception only for exceptional things (e.g. OOM) and return false for bad proof (either bad verification, or non-canonical group element). Return true for valid proof.

3) Simply return true or false, where true means that everything went fine. False means a variety of issues and they must be handled all the same.

I don't know enough java or teku to figure out which one is better, so some help would be appreciated.

---

Finally one logistical question (/cc @jtraglia ):

Should I wait for #155 to get merged before I continue here, since a non-trivial amount of work is spent on updating the ASCII tests. Or should we merge #155 after this one?

**Comment by StefanBratanov** [comment]
> Pushed some more work here.
> 
> I have a question about the Java bindings (/cc @StefanBratanov @benjaminion):
> 
> The java bindings right now, upon receiving a bad KZG proof, might either throw an exception (when we go OOM, or when a group element is not canonical) or return false (if the verification equation fails). This is a bit weird because an adversary can send a non-canonical group element or a non-verifying proof with the same ease, and they will be handled differently.
> 
> With this PR we can simplify this logic. What would you prefer to do here? Here are some options:
> 
> 1. Make all the verification functions have no retval. If they finish without any exceptions, then the proof is totally valid. Exceptions can be fine-grained but will be handled by clients.
> 2. Throw exception only for exceptional things (e.g. OOM) and return false for bad proof (either bad verification, or non-canonical group element). Return true for valid proof.
> 3. Simply return true or false, where true means that everything went fine. False means a variety of issues and they must be handled all the same.
> 
> I don't know enough java or teku to figure out which one is better, so some help would be appreciated.
> 
> Finally one logistical question (/cc @jtraglia ):
> 
> Should I wait for #155 to get merged before I continue here, since a non-trivial amount of work is spent on updating the ASCII tests. Or should we merge #155 after this one? In that case, we can just return false.

I like 3 and the current java interface won't change now which is a plus. Otherwise we need to interpret the error in Teku and check if it is actual issue or an invalid proof. My understanding is that C_KZG_BAD_VERIFY means basically proof is not valid right?  In that case, we can return false. We return true if C_KZG_OK is returned and throw an exception in all other cases as we do now throughout the java bindings.

**Comment by asn-d6** [comment]
Pushed work on Java bindings as well based on discussion with @StefanBratanov . Plus it seems to be following the stdlibs convention https://docs.oracle.com/javase/7/docs/api/java/security/Signature.html#verify(byte[])

**Comment by asn-d6** [comment]
Went over all the bindings, but did not update the test ASCII parsing, since we haven't decided what we should do with this.

**Comment by jtraglia** [comment]
> Should I wait for #155 to get merged before I continue here, since a non-trivial amount of work is spent on updating the ASCII tests. Or should we merge #155 after this one?

I think we should merge #155 first.

**Comment by StefanBratanov** [comment]
> Pushed work on Java bindings as well based on discussion with @StefanBratanov . Plus it seems to be following the stdlibs convention [https://docs.oracle.com/javase/7/docs/api/java/security/Signature.html#verify(byte[])](https://docs.oracle.com/javase/7/docs/api/java/security/Signature.html#verify(byte%5B%5D))

By the way nice implementation, thanks, didn't know JNI_TRUE/JNI_FALSE existed. Makes it cleaner.

**Comment by jtraglia** [comment]
> > Should I wait for #155 to get merged before I continue here, since a non-trivial amount of work is spent on updating the ASCII tests. Or should we merge #155 after this one?
> 
> I think we should merge #155 first.

I'm not so sure anymore. Seems that I'll need to update that PR to work with yaml tests instead of json, probably with a new PR. So nevermind, I think we should complete/review/merge this PR first.

**Comment by asn-d6** [comment]
I think this can be reviewed now!

Summary of changes to each of the bindings:
- Rust bindings return `Ok(())` if verification worked, or an appropriate `Error` otherwise (as part of a `Result`) (cc @pawanjay176)
- Java bindings maintain the same semantics: Returns `true` if verification went well, `false` if verification equation failed, and throws an exception at any other event (including non-canonical elements) as discussed above https://github.com/ethereum/c-kzg-4844/pull/158#issuecomment-1448122392 (cc @StefanBratanov)
- Python bindings don't return any retval for verification functions and just throw an exception if verification fails. (see here for an [example of a similar API](https://signify.readthedocs.io/en/latest/certificates.html#signify.x509.VerificationContext.verify)).
- I left the Node bindings with the same semantics as before (identical to Java semantics above). But they can be simplified if they follow the logic of the Python bindings below. (cc @matthewkeil)
- C# bindings now just use `C_KZG_RET` directly (cc @flcl42)
- Golang bindings return only the `CKzgRet` now and it should be checked by its caller (like before)

I also CCed the various maintainers in this message to make sure they notice this subtle change of API. I made each binding change on a separate commit to make reading easier.

File: bindings/java/c_kzg_4844_jni.c
Comment (StefanBratanov) [inline bindings/java/c_kzg_4844_jni.c:372]: Same as above. Should be `verifyBlobKzgProofBatch`

File: src/c_kzg_4844.h
Comment (jtraglia) [reply src/c_kzg_4844.h:100]: Agreed, that makes sense. Let's do this in another PR.

File: bindings/python/ckzg.c
Comment (jtraglia) [inline bindings/python/ckzg.c:135]: This formatting looks a bit strange. Is this intentional?

**Review by jtraglia — COMMENTED** [review]
Just a few nits:

File: bindings/java/c_kzg_4844_jni.c
Comment (asn-d6) [reply bindings/java/c_kzg_4844_jni.c]: Fixed in 07bb381 !

File: bindings/java/c_kzg_4844_jni.c
Comment (asn-d6) [reply bindings/java/c_kzg_4844_jni.c]: Fixed in 07bb381 !

File: bindings/csharp/Ckzg.Bindings/Ckzg.cs
Comment (asn-d6) [reply bindings/csharp/Ckzg.Bindings/Ckzg.cs:107]: Fixed in e7ea509 !

File: bindings/python/ckzg.c
Comment (asn-d6) [reply bindings/python/ckzg.c]: Fixed in 1400821 !

**Review by pawanjay176 — APPROVED** [review]
Rust bindings lgtm

**Review by jtraglia — APPROVED** [review]
Thanks for making those changes. LGTM 🥇

**Comment by matthewkeil** [comment]
> * I left the Node bindings with the same semantics as before (identical to Java semantics above). But they can be simplified if they follow the logic of the Python bindings below. (cc @matthewkeil)
 
@asn-d6  I mocked up a couple of examples for returning bool vs void.  What are your thoughts?  I tend to lean toward the bool value like the java bindings so that conditional is implicit in the function call, but honestly its not a strong opinion.  @dapplion would you like to weigh in?

```typescript
  const valid = verifyBlobKzgProof(blob, commitment, proof);
  // non-validation errors would percolate up to the outer context catch to be handled as a 500-esque error
  if (valid) {
    throw new Error(`throw an error with ${some_meta} from calling scope`);
  }
//
// --- vs ---
//
  try {
    verifyBlobKzgProof(blob, commitment, proof);
  } catch (err) {
    if (err.message.includes("C_KZG_BAD_VERIFY")) { // or similar to separate out validation vs server errors
      throw new Error(`throw an error with ${some_meta} from calling scope`);
    }
    throw err;
  }
```

**Comment by asn-d6** [comment]
After discussion with the client devs, it seems like C#, node and Java prefer the old interface of `return true for correct verification, return false for bad equation, throw exception for non-canonical element/oom`. For this reason, let's scratch this PR as it would simplify the base kzg layer, but then the bindings would get more complicated. Let's revisit if needed.

Thanks everyone for the feedback!

**Pull request closed without merging**
